### PR TITLE
OLMIS-730: Reference Data should be Immutable

### PR DESCRIPTION
**Pull request by benleibert**
Limit our reference-service’s API primarily to GET operations. It may optionally support object creation, but not modification, for now.

The UserRepository and UserController classes exist as placeholders for how other entities may potentially be exposed once we determine what their endpoints should be.

---
This change is [Reviewable](https://reviewable.io/reviews/openlmis/openlmis-requisition/1)

### CODE REVIEW

File: src/main/java/org/openlmis/referencedata/web/UserController.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+    @Autowired
+    private UserRepository userRepository;
+
+    /*
+        Partially due to story OLMIS-726, which involves replacing id fileds with non-sequential UUID values,
+        this POST endpoint will likely change. It exists in its current form largely as a placeholder.
+     */
+    @RequestMapping(path = "/api/users/", method = RequestMethod.POST)
+    public void createUser(@RequestBody User user)
```
Comment (joshzamor) [inline src/main/java/org/openlmis/referencedata/web/UserController.java]: OLMIS-730 #comment Hey Ben, what does this method do?  If I post {"id": "123"} I will get a 405, method not allowed, I assume due to the save(User) method not being published on the Spring Data repository.  (which happens to be very useful to know as I wondered what effect that would have on non-generated REST interfaces).

**Comment by joshzamor** [comment]
@jkondrat , please take a look at this pull request.  This branch is now breaking as it used a Sping Data JPA Repository that wasn't CrudRepository which doesn't work well with the integration testing setup now in master.  With your integration testing setup in #5 it seems like the generic testing assumption that every repository would be a CrudRepository is perhaps putting a constraint on which repository design we use.  Please take a look and reconsider the CrudRepository assumption.

**Comment by benleibert** [comment]
Hi @joshzamor,

Because you pointed out that the latest version of this branch is broken, I re-tested POSTs to /api/users/ within the context of commit-version 0a0264d. 

As the comment above the method advertises, not a lot of thought was given to the “id.” Looking at it again, a little more should have been. Specifically, the method is intended to allow clients to POST a new object. As such, there’s no need for it to require that an “id” field be provided, and the fact that it does is confusing. That being said, if clients supply an “id” along with all User fields marked as non-nullable, the  method works as intended. An example of a request-body that works for me is:

{
 "id": "1",
 
  "username": "user1",
  "firstName": "first1",
  "lastName": "last1",
  "password" : "password",
  "verified": true,
  "active": true
}

A request-body of {"id": "123"} fails because it’s missing non-nullable fields. 

As an aside, the response codes returned by the method need to be improved. As the comment above createUser suggests in a round-about way, it’s really just there to highlight that the save(User) method on the UserRepository exists and works. The Save(User) method is indirectly created within the repository by the last few lines of UserRepository.java.
 
Thank you,
Ben

**Comment by joshzamor** [comment]
@benleibert , Regarding the createUser method in the User controller, I don't have that working with commit 0a0264dfee8051d878cd884b21e2cf76dce77b4c.  When I POST to /api/users your JSON above with content type application/json I get a 405 method not allowed.  There are a couple questions going on, for this one I was just curious if unpublishing the save() method on the repository with @RestResource had any effect outside the RESTful interface.

<details><summary><i>Previously, benleibert wrote…</i></summary>
> Hi @joshzamor,
> 
> 
> 
> Because you pointed out that the latest version of this branch is broken, I re-tested POSTs to /api/users/ within the context of commit-version 0a0264d. 
> 
> 
> 
> As the comment above the method advertises, not a lot of thought was given to the “id.” Looking at it again, a little more should have been. Specifically, the method is intended to allow clients to POST a new object. As such, there’s no need for it to require that an “id” field be provided, and the fact that it does is confusing. That being said, if clients supply an “id” along with all User fields marked as non-nullable, the  method works as intended. An example of a request-body that works for me is:
> 
> 
> 
> {
> 
>  "id": "1",
> 
>  
> 
>   "username": "user1",
> 
>   "firstName": "first1",
> 
>   "lastName": "last1",
> 
>   "password" : "password",
> 
>   "verified": true,
> 
>   "active": true
> 
> }
> 
> 
> 
> A request-body of {"id": "123"} fails because it’s missing non-nullable fields. 
> 
> 
> 
> As an aside, the response codes returned by the method need to be improved. As the comment above createUser suggests in a round-about way, it’s really just there to highlight that the save(User) method on the UserRepository exists and works. The Save(User) method is indirectly created within the repository by the last few lines of UserRepository.java.
> 
>  
> 
> Thank you,
> 
> Ben
</details>


---







*Comments from [Reviewable](https://reviewable.io:443/reviews/openlmis/openlmis-requisition/1#-:-KJpqcj3bwukM1hK56V5:986483810)*

**Comment by benleibert** [comment]
Hi Josh,

It looks like you need to POST to “/api/users/” (with a trailing slash) rather than “/api/users.” If you try that, I think you’ll see the behavior I had described.

Thank you,
Ben

**Comment by benleibert** [comment]
Hi Josh,

It looks like you need to POST to “/api/users/” (with a trailing slash) rather than “/api/users.” If you try that, I think you’ll see the behavior I had described.

Thank you,
Ben

**Comment by joshzamor** [comment]
:lgtm:

<details><summary><i>Previously, benleibert wrote…</i></summary>
> Hi Josh,
> 
> 
> 
> It looks like you need to POST to “/api/users/” (with a trailing slash) rather than “/api/users.” If you try that, I think you’ll see the behavior I had described.
> 
> 
> 
> Thank you,
> 
> Ben
</details>


---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/openlmis/openlmis-requisition/1#-KJvKMVUfzafE_i9EC6a:-KJvKMVUfzafE_i9EC6b:-653792965) (no related file):*
Last piece:  integration tests are failing.  Thanks Ben!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openlmis/openlmis-requisition/1#-:-KJvKKWConxR7DcyXMZj:1417088721)*

**Comment by benleibert** [comment]
*[a discussion](https://reviewable.io:443/reviews/openlmis/openlmis-requisition/1#-KJvKMVUfzafE_i9EC6a:-KK577-o4PFY-phZHXVd:2039813319) (no related file):*
<details><summary><i>Previously, joshzamor (Josh Zamor) wrote…</i></summary>
> Last piece:  integration tests are failing.  Thanks Ben!
</details>
Thanks Josh - I addressed them.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openlmis/openlmis-requisition/1)*

**Pull request merged**